### PR TITLE
CSF: Rename `preview.js` `globals` to `initialGlobals`

### PR DESCRIPTION
**Pull request by shilman**
Closes N/A

## What I did

CSF updates to implement Story globals per https://github.com/storybookjs/storybook/discussions/23347#discussioncomment-9575546

- [x] Add preview.js initialGlobals to replace globals
- [x] Deprecate preview.js globals
- [x] Add globals to component/story annotations
- [x] Add automigration
- [x] Add MIGRATION.md notes
- [x] Bugfix for `ConfigFile.removeField` common case

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

0. Build this branch
1. Create a sandbox or a random storybook project
2. Add a `globals` annotation to `preview.js`
3. Verify that the initial value of globals are set correctly and a deprecation warning is shown
4. In the project directory run: `/path/to/code/lib/cli/bin/index.js automigrate initial-globals`
5. Verify that the prompts make sense and that `preview.js` has been updated to `initialGlobals`
6. Verify that the initial value of globals are still set correctly

### Documentation

Interestingly, could not find `globals` documented on https://storybook.js.org/docs/get-started cc @kylegach 

- [ ] Add or update documentation reflecting your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/cli/src/automigrate/fixes/initial-globals.test.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  await expect(check(`export default { tags: ['a', 'b']}`)).resolves.toBeFalsy();
+});
+
+it('initialGlobals setting', async () => {
+  await expect(check(`export default { initialGlobals: { a:  1 } }`)).resolves.toBeFalsy();
+});
+
+it('globals setting', async () => {
+  await expect(check(`export default { globals: { a:  1 } }`)).resolves.toBeTruthy();
+});
```
Comment (tmeasday) [inline code/lib/cli/src/automigrate/fixes/initial-globals.test.ts:33]: What about non-default exports?

File: code/lib/preview-api/src/modules/store/StoryStore.ts [hunk 2]
```diff
@@ -77,10 +77,10 @@ export class StoryStore<TRenderer extends Renderer> {
     this.storyIndex = new StoryIndexStore(storyIndex);
 
     this.projectAnnotations = normalizeProjectAnnotations(projectAnnotations);
-    const { globals, globalTypes } = projectAnnotations;
+    const { globals, initialGlobals, globalTypes } = projectAnnotations;
 
     this.args = new ArgsStore();
-    this.globals = new GlobalsStore({ globals, globalTypes });
+    this.globals = new GlobalsStore({ globals: initialGlobals ?? globals, globalTypes });
```
Comment (tmeasday) [inline code/lib/preview-api/src/modules/store/StoryStore.ts:83]: I think this bit should be part of the `normalizeProjectAnnotations` call.

Also it should give a deprecation warning.

**Review by tmeasday — CHANGES_REQUESTED** [review]

File: code/lib/cli/src/automigrate/fixes/initial-globals.test.ts (hunk 1 above)
Comment (shilman) [reply code/lib/cli/src/automigrate/fixes/initial-globals.test.ts:33]: This is already handled by `ConfigFile` unit tests

Comment (tmeasday) [reply code/lib/cli/src/automigrate/fixes/initial-globals.test.ts:33]: I'm not sure I understand @shilman. If the user has `export const globals = {...}` what happens?

Comment (shilman) [reply code/lib/cli/src/automigrate/fixes/initial-globals.test.ts:33]: `export const globals = { ... }` will get renamed to `export const initialGlobals = { .. }` (just hand-tested to verify).

The `ConfigFile` class takes care of three different config file syntaxes:
1. `export default { globals: { ... } }`
2. `export const globals = { ... }`
3. `module.exports = { globals: { } }` 

All of these cases are tested in the `ConfigFile` tests, so I didn't feel it was necessary to repeat those tests here. See https://github.com/storybookjs/storybook/blob/next/code/lib/csf-tools/src/ConfigFile.test.ts

**Review by tmeasday — CHANGES_REQUESTED** [review]
Can I get a test on `normalizeProjectAnnotations` please?

File: code/lib/preview-api/src/modules/store/csf/composeConfigs.ts [hunk 3]
```diff
@@ -42,6 +42,10 @@ export function composeConfigs<TRenderer extends Renderer>(
 ): ProjectAnnotations<TRenderer> {
   const allArgTypeEnhancers = getArrayField(moduleExportList, 'argTypesEnhancers');
   const stepRunners = getField(moduleExportList, 'runStep');
+  const initialGlobals = combineParameters(
+    getObjectField(moduleExportList, 'globals'),
+    getObjectField(moduleExportList, 'initialGlobals')
+  );
```
Comment (tmeasday) [inline code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:48]: Shouldn't all configs be normalized before reaching `composeConfigs`?

File: code/lib/preview-api/src/modules/store/csf/composeConfigs.ts [hunk 4]
```diff
@@ -55,7 +59,8 @@ export function composeConfigs<TRenderer extends Renderer>(
       ...allArgTypeEnhancers.filter((e) => !e.secondPass),
       ...allArgTypeEnhancers.filter((e) => e.secondPass),
     ],
-    globals: getObjectField(moduleExportList, 'globals'),
+    globals: initialGlobals, // deprecated
```
Comment (tmeasday) [inline code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:62]: What is this used for? This seems confusing/wrong?

Comment (shilman) [reply code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:62]: Technically speaking this is a breaking change to portable stories if we remove it.

File: code/lib/preview-api/src/modules/store/csf/composeConfigs.ts (hunk 3 above)
Comment (shilman) [reply code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:48]: Good call

File: code/lib/preview-api/src/modules/store/csf/composeConfigs.ts (hunk 4 above)
Comment (tmeasday) [reply code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:62]: Why?

Comment (shilman) [reply code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:62]: If some user code calls `composeConfigs` and expects `globals` to be set, and suddenly it is not set, that's a breaking change

Comment (tmeasday) [reply code/lib/preview-api/src/modules/store/csf/composeConfigs.ts:62]: OK

**Comment by shilman** [comment]
Self-merging @tmeasday

**Pull request merged**